static/app/views/releases/thresholdsList/thresholdGroupRows.tsx

```diff
@@ -0,0 +1,413 @@
+    [key: string]: EditingThreshold;
+  }>({});
+  const [newThresholdIterator, setNewThresholdIterator] = useState<number>(0); // used simply to initialize new threshold
+  const api = useApi();
+  const project = thresholds[0].project;
+  const environment = thresholds[0].environment;
+  const defaultWindow = thresholds[0].window_in_seconds;
+
+  const thresholdsById: {[id: string]: Threshold} = useMemo(() => {
+    const byId = {};
+    thresholds.forEach(threshold => {
+      byId[threshold.id] = threshold;
+    });
+    return byId;
+  }, [thresholds]);
+
+  const thresholdIdSet = useMemo(() => {
+    return new Set([
+      ...thresholds.map(threshold => threshold.id),
+      ...Object.keys(editingThresholds),
+    ]);
+  }, [thresholds, editingThresholds]);
+
+  const initializeNewThreshold = () => {
+    const thresholdId = `${NEW_THRESHOLD_PREFIX}-${newThresholdIterator}`;
+    const [windowValue, windowSuffix] = parseLargestSuffix(defaultWindow);
+    const newThreshold: EditingThreshold = {
+      id: thresholdId,
+      project,
+      environment,
+      windowValue,
+      windowSuffix,
+      threshold_type: 'total_error_count',
+      trigger_type: 'over',
+      value: 0,
+      hasError: false,
+    };
+    const updatedEditingThresholds = {...editingThresholds};
+    updatedEditingThresholds[thresholdId] = newThreshold;
+    setEditingThresholds(updatedEditingThresholds);
+    setNewThresholdIterator(newThresholdIterator + 1);
+  };
+
+  const enableEditThreshold = thresholdId => {
+    const updatedEditingThresholds = {...editingThresholds};
+    const threshold = JSON.parse(JSON.stringify(thresholdsById[thresholdId]));
+    const [windowValue, windowSuffix] = parseLargestSuffix(threshold.window_in_seconds);
+    updatedEditingThresholds[thresholdId] = {
+      ...threshold,
+      windowValue,
+      windowSuffix,
+      hasError: false,
+    };
+    setEditingThresholds(updatedEditingThresholds);
+  };
+
+  const saveThreshold = thresholdIds => {
+    thresholdIds.forEach(id => {
+      const thresholdData = editingThresholds[id];
+      const submitData: {[key: string]: any} = {...thresholdData};
+      let path = `/projects/${orgSlug}/${thresholdData.project.slug}/release-thresholds/${id}/`;
+      let method: APIRequestMethod = 'PUT';
+      if (id.includes(NEW_THRESHOLD_PREFIX)) {
+        path = `/projects/${orgSlug}/${thresholdData.project.slug}/release-thresholds/`;
+        method = 'POST';
+      }
+      const seconds = moment
+        // @ts-ignore: ts is unhappy with moment string duration format
+        .duration(thresholdData.windowValue, thresholdData.windowSuffix)
+        .as('seconds');
+      submitData.window_in_seconds = seconds;
+      submitData.environment = thresholdData.environment.name; // api expects environment as a string
+      const request = api.requestPromise(path, {
+        method,
+        data: submitData,
+      });
+      request
+        .then(() => {
+          refetch();
```

Comment:
refetch might override things if they've made additional edits, just something to think about

Reply:
Yup, already considered that!
Since the "in flight edits" are already set in state - their form data actually won't be affected by the refetch :)